RecoBTag/FeatureTools/plugins/HiggsInteractionNetTagInfoProducer.cc

```diff
@@ -0,0 +1,353 @@
+    edm::RefToBase<reco::Jet> jet_ref(jets, jet_n);
+
+    // create jet features
+    DeepBoostedJetFeatures features;
+    // declare all the feature variables (init as empty vector)
+    for (const auto &name : cpf_features_) {
+      features.add(name);
+    }
+    for (const auto &name : sv_features_) {
+      features.add(name);
+    }
+
+    // fill values only if above pt threshold and has daughters, otherwise left
+    // empty
+    bool fill_vars = true;
+    if (jet.pt() < min_jet_pt_)
+      fill_vars = false;
+    if (jet.numberOfDaughters() == 0)
+      fill_vars = false;
+
+    if (fill_vars) {
+      fillChargedParticleFeatures(features, jet);
+      fillSVFeatures(features, jet);
+
+      features.check_consistency(cpf_features_);
+      features.check_consistency(sv_features_);
+    }
+
+    // this should always be done even if features are not filled
+    output_tag_infos->emplace_back(features, jet_ref);
+  }
+
+  iEvent.put(std::move(output_tag_infos));
+}
+
+void HiggsInteractionNetTagInfoProducer::fillChargedParticleFeatures(DeepBoostedJetFeatures &fts,
+                                                                     const reco::Jet &jet) {
+  // require the input to be a pat::Jet
+  const auto *patJet = dynamic_cast<const pat::Jet *>(&jet);
+  if (!patJet) {
+    throw edm::Exception(edm::errors::InvalidReference) << "Input is not a pat::Jet.";
+  }
+
+  // do nothing if jet does not have constituents
+  if (jet.numberOfDaughters() == 0)
+    return;
+
+  // some jet properties
+  math::XYZVector jet_dir = jet.momentum().Unit();
+  GlobalVector jet_ref_track_dir(jet.px(), jet.py(), jet.pz());
+  const float etasign = jet.eta() > 0 ? 1 : -1;
+
+  std::vector<reco::CandidatePtr> daughters;
+  for (const auto &cand : jet.daughterPtrVector()) {
+    // get the original reco/packed candidate
+    auto daugh = pfcands_->ptrAt(cand.key());
+    if (daugh->charge() != 0 && daugh->pt() > min_pt_for_track_properties_)
+      daughters.push_back(daugh);
+  }
+
+  // reserve space
+  for (const auto &name : cpf_features_) {
+    fts.reserve(name, daughters.size());
+  }
+
+  auto useTrackProperties = [&](const reco::PFCandidate *reco_cand) {
+    const auto *trk = reco_cand->bestTrack();
+    return trk != nullptr && trk->pt() > min_pt_for_track_properties_;
+  };
+
+  // sort charged pf candidates by 2d impact parameter significance
+  std::vector<btagbtvdeep::SortingClass<reco::CandidatePtr>> c_sorted;
+  for (const auto &cand : daughters) {
+    TrackInfoBuilder trkinfo(track_builder_);
+    trkinfo.buildTrackInfo(&(*cand), jet_dir, jet_ref_track_dir, *pv_);
+    c_sorted.emplace_back(cand,
+                          trkinfo.getTrackSip2dSig(),
+                          -btagbtvdeep::mindrsvpfcand(*svs_, &(*cand), jet_radius_),
+                          cand->pt() / jet.pt());
+  }
+
+  std::sort(c_sorted.begin(), c_sorted.end(), btagbtvdeep::SortingClass<reco::CandidatePtr>::compareByABCInv);
```

Comment:
The only issue I noticed is that when there are more than 60 charged PF candidates (which happens rarely but not super rarely), then since we only take the first 60 in the tag producer, we are actually taking a different set than the ones used to train the algorithm.